### PR TITLE
REPLAY-1517 Add unsupported views recorder

### PR DESCRIPTION
**Pull request by maciejburda**
### What and why?

Adds special recorder that recognises unsupported views and provides a common placeholder for them.

Snapshot:
testUnsupportedView()-allowAll-privacy

### How?

A brief description of implementation details of this PR.

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [ ] Run unit tests
- [ ] Run integration tests
- [ ] Run smoke tests

### CODE REVIEW

**Comment by maciejburda** [comment]
Add snapshot tests

**Review by ncreated — CHANGES_REQUESTED** [review]
Looks ok 👍 with few minor remarks on unit tests. Blocking until we add snapshots.

File: DatadogSessionReplay/SRSnapshotTests/SRSnapshotTests/SRSnapshotTests.swift [hunk 1]
```diff
@@ -203,4 +203,16 @@ final class SRSnapshotTests: SnapshotTestCase {
...
         }
     }
+
+    func testUnsupportedView() throws {
+        show(fixture: .unsupportedViews)
+
+        let image = try takeSnapshot(configuration: .init(privacy: .allowAll))
+        DDAssertSnapshotTest(
+            newImage: image,
+
```
Comment (ncreated) [inline DatadogSessionReplay/SRSnapshotTests/SRSnapshotTests/SRSnapshotTests.swift:213]:
```suggestion
```

File: DatadogSessionReplay/SRSnapshotTests/SRHost/Fixtures/UnsupportedViewsViewController.swift [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+        addChild(hostingController)
+        swiftUIContainer.addSubview(hostingController.view)
+        hostingController.view.frame = swiftUIContainer.bounds
+        hostingController.didMove(toParent: self)
+    }
+}
+
+import SwiftUI
+
+struct TestView: View {
```
Comment (ncreated) [inline DatadogSessionReplay/SRSnapshotTests/SRHost/Fixtures/UnsupportedViewsViewController.swift:26]: nit/ It's a generic name and will be confusing in internal namespace. Because we use it only in this file:
```suggestion
fileprivate struct TestView: View {
```

File: DatadogSessionReplay/SRSnapshotTests/SRSnapshotTests.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -26,7 +26,8 @@
 		61E7DFB9299A5A3E001D7A3A /* Basic.storyboard in Resources */ = {isa = PBXBuildFile; fileRef = 61E7DFB8299A5A3E001D7A3A /* Basic.storyboard */; };
 		61E7DFBB299A5C9D001D7A3A /* BasicViewControllers.swift in Sources */ = {isa = PBXBuildFile; fileRef = 61E7DFBA299A5C9D001D7A3A /* BasicViewControllers.swift */; };
 		61EC1A37299CFD7E00224FB6 /* TestUtilities in Frameworks */ = {isa = PBXBuildFile; productRef = 61EC1A36299CFD7E00224FB6 /* TestUtilities */; };
-		A77C62F429E4298200AC923B /* ImagesViewControllers.swift in Sources */ = {isa = PBXBuildFile; fileRef = A77C62F329E4298200AC923B /* ImagesViewControllers.swift */; };
```
Comment (ncreated) [inline DatadogSessionReplay/SRSnapshotTests/SRSnapshotTests.xcodeproj/project.pbxproj:29]: blocking/ When running all snapshot tests, the `testImages()` crashes:
```
Test Case '-[SRSnapshotTests.SRSnapshotTests testImages]' started.
2023-04-13 12:36:37.480228+0200 SRHost[27731:16412801] [Storyboard] Unknown class _TtC6SRHost20ImagesViewController in Interface Builder file.
2023-04-13 12:36:37.516184+0200 SRHost[27731:16412801] *** Terminating app due to uncaught exception 'NSUnknownKeyException', reason: '[<UIViewController 0x12b72f6f0> setValue:forUndefinedKey:]: this class is not key value coding-compliant for the key customButton.'
*** First throw call stack:
(
	0   CoreFoundation                      0x000000018040e7c8 __exceptionPreprocess + 172
	1   libobjc.A.dylib                     0x0000000180051144 objc_exception_throw + 56
```

It might be coming from this unnecessary change (removal of `ImagesViewControllers.swift`).

File: DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/UnsupportedViewRecorder.swift [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+fileprivate extension String {
+    func extractGenericClassNameIfNeeded() -> String {
+        let regex = try? NSRegularExpression(pattern: "<([^>]*)>[^<>]*$", options: [])
+        if let match = regex?.firstMatch(in: self, range: NSRange(startIndex..., in: self)),
+            let range = Range(match.range(at: 1), in: self) {
+            return String(self[range])
+        }
+        return self
+    }
```
Comment (ncreated) [inline DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/UnsupportedViewRecorder.swift:73]: Running in on the main thread (from `semantics(of:with:in:)` sounds bad idea. It makes the call 2 times slower (`~0.005ms` without → `~0.012ms` with regex). Given that we want to be performance-first and this logic is fairly easy to move to background thread (to `buildWireframes(with:)`), we should avoid spending time on the main queue if possible.

On top of that, if we want to enter regexps world, we need a fair test coverage for it (not only happy path).

On top of on top 🙂 - what does it do exactly? Is it critical 🤔? It needs some code comments (the general price for most write-only regexps 😝), otherwise it's too ambiguous.

**Review by ncreated — CHANGES_REQUESTED** [review]
I'm happy we're adding "unsupported" visuals 🙂! 

I left few feedbacks - some are blocking.

File: DatadogSessionReplay/SRSnapshotTests/SRSnapshotTests.xcodeproj/project.pbxproj (hunk 3 above)
Comment (maciejburda) [reply DatadogSessionReplay/SRSnapshotTests/SRSnapshotTests.xcodeproj/project.pbxproj:29]: It's been a rebase tweak I thought I resolved. Thanks for catching!

File: DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/UnsupportedViewRecorder.swift (hunk 4 above)
Comment (maciejburda) [reply DatadogSessionReplay/Sources/Recorder/ViewTreeSnapshotProducer/ViewTreeSnapshot/NodeRecorders/UnsupportedViewRecorder.swift:73]: Yeah - I was on a fence with this. It's removes `SwiftUI._UIHostingView<>` from the class name and keeps what is between <>.

I'll remove it and we can reconsider it at later stage.

**Review by ncreated — APPROVED** [review]
💪🏅
